ml-agents-envs/mlagents/envs/sampler_class.py

```diff
@@ -0,0 +1,112 @@
+                )
+
+    def sample_parameter(self):
+        return np.random.uniform(self.min_value, self.max_value)
+    
+
+class MultiRangeUniformSampler(Sampler):
+    def __init__(self, param_dict):
+        self.check_keys(param_dict)
+        self.intervals = param_dict["intervals"]
+    
+    def check_keys(self, param_dict):
+        keys = ["intervals"]
+        for key in keys:
+            if key not in param_dict:
+                raise SamplerException(
+                    key + " is required to use MultiRange Uniform sampler but" 
+                    " this value isn't provided. Please recheck"
+                    " the documentation to ensure that the sampler file"
+                    " matches the required format."
+                )
+
+    
+    def sample_parameter(self):
+        cur_min, cur_max = self.intervals[np.random.randint(len(self.intervals))]
+        return np.random.uniform(cur_min, cur_max)
+
+
+class GaussianSampler(Sampler):
+    def __init__(self, param_dict):
+        self.check_keys(param_dict)
+        self.mean = param_dict["mean"]
+        self.var = param_dict["var"]
+    
+    def sample_parameter(self):
+        return np.random.normal(self.mean, self.var)
+
+    def check_keys(self, param_dict):
+        keys = ["mean", "var"]
+        for key in keys:
+            if key not in param_dict:
+                raise SamplerException(
+                    key + " is required to use MultiRange Uniform sampler but" 
+                    " this value isn't provided. Please recheck"
+                    " the documentation to ensure that the sampler file"
+                    " matches the required format."
+                )
+
+
+
+# To introduce new sampling methods, just need to 'register' them to this sampler factory
+class SamplerFactory:
+    NAME_TO_CLASS = {
+    "uniform": UniformSampler,
+    "gaussian": GaussianSampler,
+    "multirange_uniform": MultiRangeUniformSampler,
+    }
+
+    @staticmethod
+    def register_sampler(name, sampler_cls):
+        SamplerFactory.NAME_TO_CLASS[name] = sampler_cls
+    
+    def init_sampler_class(name, param_dict):
+        sampler_cls = SamplerFactory.NAME_TO_CLASS[name]
+        return sampler_cls(param_dict)
+
+
+class SamplerManager:
+    def __init__(self, reset_param_dict):
+        self.reset_param_dict = reset_param_dict
+        self.sampler_factory = SamplerFactory()
+        self.samplers = {}
+        for param_name, cur_param_dict in self.reset_param_dict.items():
+            # cur_param_dict = reset_param_dict[param_name]
+            sampler_name = cur_param_dict["sampler-type"]
+            param_sampler = SamplerFactory.init_sampler_class(sampler_name, cur_param_dict)
+
+            self.samplers[param_name] = param_sampler
+
+    def sample_all(self, train_mode = True):
+        res = {}
+        for param_name, param_sampler in self.samplers.items():
+            res[param_name] = param_sampler.sample_parameter()
+        return res
```

Comment:
nit: newline at EOF. if you're using pycharm, you can use this to make it automatic 
![image](https://user-images.githubusercontent.com/6877802/60537479-78be4c00-9cbd-11e9-9c7e-40d96e2d1e22.png)

Reply:
Also: use Black to format your code, it will do this automatically as well: https://black.readthedocs.io/en/stable/

Reply:
So I tried setting np.random's seed, but there isn't a deterministic selection for the reset parameter sampling. I am guessing this is happening because in sample_all you can call the different sampling methods in a different order in each loop, which changes subsequent values. 

One alternative could be to maintain a seed in each of the sampler methods (i.e. - maintain the seed in each instance of UniformSampler or other sampling classes, though this would require a lot of calls to np.random.get_state to maintain the next seed).

Another idea is to deterministically loop over the self.samplers.items().

What are your thoughts?